core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -107,42 +107,86 @@ class LogSegment private[log] (val log: FileRecords,
+   *                                 for KAFKA-5413. Typically, the compactor might encounter such log segments.
+   */
+  @nonthreadsafe
+  private[log] def append(largestOffset: Long,
+                          largestTimestamp: Long,
+                          shallowOffsetOfMaxTimestamp: Long,
+                          records: MemoryRecords,
+                          allowOversizeIndexOffset: Boolean): Unit = {
     if (records.sizeInBytes > 0) {
       trace(s"Inserting ${records.sizeInBytes} bytes at end offset $largestOffset at position ${log.sizeInBytes} " +
             s"with largest timestamp $largestTimestamp at shallow offset $shallowOffsetOfMaxTimestamp")
       val physicalPosition = log.sizeInBytes()
+      var canAppendToIndex = true
+
       if (physicalPosition == 0)
         rollingBasedTimestamp = Some(largestTimestamp)
+
+      // Must be able to convert the largest offset to base-relative offset.
+      // Because of KAFKA-5413, a log segment could contain messages larger than firstOffset + Integer.MAX_VALUE. If
+      // that is the case, silently ignore the fact that we cannot convert all message offsets to the base-relative
+      // form. Also make sure we do not generate an index entry for such messages, so that there is no further issue
+      // downstream.
+      if (!canConvertToRelativeOffset(largestOffset)) {
+        if (allowOversizeIndexOffset)
+          canAppendToIndex = false
+        else
+          throw new IllegalArgumentException("requirement failed: " +
+                                             "largest offset in message set can not be safely " +
+                                             "converted to relative offset.")
+      }
+
       // append the messages
-      require(canConvertToRelativeOffset(largestOffset), "largest offset in message set can not be safely converted to relative offset.")
       val appendedBytes = log.append(records)
       trace(s"Appended $appendedBytes to ${log.file()} at end offset $largestOffset")
       // Update the in memory max timestamp and corresponding offset.
       if (largestTimestamp > maxTimestampSoFar) {
         maxTimestampSoFar = largestTimestamp
         offsetOfMaxTimestamp = shallowOffsetOfMaxTimestamp
       }
+
       // append an entry to the index (if needed)
-      if(bytesSinceLastIndexEntry > indexIntervalBytes) {
-        offsetIndex.append(largestOffset, physicalPosition)
-        timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
-        bytesSinceLastIndexEntry = 0
+      if (bytesSinceLastIndexEntry > indexIntervalBytes) {
+        if (canAppendToIndex) {
+          offsetIndex.append(largestOffset, physicalPosition)
+          timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+          bytesSinceLastIndexEntry = 0
+        } else {
+          debug("Skipping append to offset index to prevent offset overflow")
```

Comment:
Perhaps we can add the segment filename to this log message? `log.file.getAbsolutePath`